### PR TITLE
Release 3.0

### PR DESCRIPTION
**Pull request by bitsofbits**
Make a 3.0 release that is compatible with 3.0 pipelines.

This includes the changes from an earlier, never merged PR:

* Switch from using `slim` API, which is quasi deprecated, to the new, `estimators`/`datasets` API
* Remove old, unused files.
* Refactor and simplify some of the over engineered bits from original implementation.
      - Break up `utility.py` into manageable chunks.
* Clean up add-hoc way we were grabbing chunks of feature data.
      - Straightforward implementation was slow (probably reason for ad-hockery to begin with).
      - Reimplement for reasonable speed.
* Update the `vessel_identitfy` training files to use `pipe_production_b`

As well as some new changes:

* More prefetching to try to further  improve training speed.
* Go back to original random rolling of data during training since tests showed slightly better
  accuracy.
* Changes to support UVI and MMSI simultaneously.
* Fix way vessel types are upsampled
* Fix vessel metrics to work with vessel_id.
* Correctly stratify data using new classes.
* Generate training data directly from Vessel Database
* Modify training invocation to make tracking runs easier.

- [x] Setup as v 3.0 release
- [x] Check fishing output
- [x] Check vessel output

Optional:

- [x] Check fishing training.
- [x] Check vessel Training

These last two have been tested as the PR was worked on, but haven't been fully tested with the final 
code on the branch.  However, vessel training in particular is slow and isn't critical for the initial
3.0 release as we have weights from earlier runs and the training portion of the repo doesn't
directly effect the rest of the 3.0 repos.

### CODE REVIEW

File: classification/feature_generation/file_iterator_test.py [hunk 1]
```diff
@@ -64,11 +46,13 @@ def log_mem():
 
 
 def test_read_files_from_gcs():
-    path = "gs://world-fishing-827/data-production/classification/release-0.1.2/pipeline/output/features/mmsis/part-00000-of-00001.txt"
+    path = "gs://world-fishing-827/data-production/classification/release-0.1.2/pipeline/output/ids/part-00000-of-00001.txt"
```
Comment (enriquetuya) [inline classification/feature_generation/file_iterator_test.py]: @bitsofbits can we move `gs://world-fishing-827/data-production/classification/release-0.1.2/` to somewhere else? something like `gfw-repositories-assets-vessel-classification/classification/release-3.0.0/`?

File: classification/feature_generation/fishing_feature_generation.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+            num_slices_per_id=8):
+
+    random_state = np.random.RandomState()
+
+    
+    def xform(id_, movement_features):
+
+        def _xform(features, int_id):
+            # Extract several random windows from each vessel track
+            # TODO: Fix feature generation so it returns strings directly
```
Comment (enriquetuya) [inline classification/feature_generation/fishing_feature_generation.py]: Should we solve the TODO?

File: classification/metrics/compute_vessel_metrics.py [hunk 3]
```diff
@@ -15,12 +15,12 @@
 
 Example:
 
+    python -m classification.metrics.compute_vessel_metrics \
+        --inference-table machine_learning_dev_ttl_120d.vessel_char_vid_features_v20190509  \
```
Comment (enriquetuya) [inline classification/metrics/compute_vessel_metrics.py]: @bitsofbits I would like to know a bit more about parameters, know which one is input, which one is output. For input, which points to where it is generated?

File: classification/metrics/compute_vessel_metrics.py [hunk 4]
```diff
@@ -769,39 +659,80 @@ def confusion_matrix(results):
...
-                # Parsing dates is expensive and all extractors use dates, so parse them
-                # once up front
-                row['start_time'] = _parse(row['start_time'])
-                #dateutil.parser.parse(row['start_time'])
-                for ext in extractors:
-                    ext.extract(row)
+    query = """
+
+    SELECT c.* except (vessel_id), vessel_id as id FROM 
+    `{}` b
```
Comment (enriquetuya) [inline classification/metrics/compute_vessel_metrics.py]: I would assign `b` and `c` better names, it is not clear without knowing what you are doing which tables you are querying.

File: classification/metrics/compute_vessel_metrics.py [hunk 5]
```diff
@@ -1073,15 +988,37 @@ def dump_html(args, results):
...
 
+# TODO: compute the fraction of each vessel type, then compute
+# the weights for precision recall, accuracy with weighted
+# values rather than equal values.
+# Query for weights looks like:
+'''
+with
+
+core as (
+select * from `machine_learning_dev_ttl_120d.vessel_char_vid_features_v20190509*`
```
Comment (enriquetuya) [inline classification/metrics/compute_vessel_metrics.py]: @bitsofbits you shouldn't be querying a table in a ttl dataset, if this needs to be re-run move it to a permanent location.

File: classification/feature_generation/fishing_feature_generation.py (hunk 2 above)
Comment (bitsofbits) [reply classification/feature_generation/fishing_feature_generation.py]: @enriquetuya : that involves enough changes that I'm reluctant to do that now.  My tentative plan is fix this on the bird life style feature generation I putting together for loitering.  Then start moving models over to that pipeline.

File: classification/feature_generation/file_iterator_test.py (hunk 1 above)
Comment (bitsofbits) [reply classification/feature_generation/file_iterator_test.py]: @enriquetuya : This is very old, so it's really only kept around to have some consistent data for testing.  Unsharded features live `pipe-production-X` and sharded features are ephemeral. Given that we can copy just a subset of these somewhere to use for testing. Do you have a preference for what it's called?

File: classification/metrics/compute_vessel_metrics.py (hunk 4 above)
Comment (bitsofbits) [reply classification/metrics/compute_vessel_metrics.py]: Done

File: classification/metrics/compute_vessel_metrics.py (hunk 5 above)
Comment (bitsofbits) [reply classification/metrics/compute_vessel_metrics.py]: This should be a passed in parameter anyway. Gone now.

File: classification/feature_generation/feature_utilities.py [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+
+    slice_length = len(slice)
+    assert (slice_length <= window_size)
+    reps = int(np.ceil(window_size / float(slice_length)))
+    return np.concatenate([slice] * reps, axis=0)[:window_size]
+
+def np_zero_pad_slice(slice, window_size, random_state):
+    """ Pads slice to the specified window size.
+
+  Series that are shorter than window_size are repeated into unfilled space.
```
Comment (d40cht) [inline classification/feature_generation/feature_utilities.py]: Comment is incorrect.

File: classification/feature_generation/feature_utilities.py [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+  Returns:
+    a numpy array of length window_size in the first dimension.
+  """
+
+    slice_length = len(slice)
+    delta = window_size - slice_length
+    assert delta >= 0
+    _, n_feat = slice.shape
+    padded = np.zeros([window_size, n_feat], dtype=slice.dtype)
+    offset = random_state.randint(0, delta + 1)
```
Comment (d40cht) [inline classification/feature_generation/feature_utilities.py]: This could do with some explanation. Presumably for data augmentation?

File: classification/feature_generation/feature_utilities.py [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+    return np.concatenate([slice] * reps, axis=0)[offset:offset+window_size]
+
+def np_pad_repeat_slice_2(slice, window_size, random_state):
+    """ Pads slice to the specified window size.
+
+  Series that are shorter than window_size are repeated into unfilled space.
```
Comment (d40cht) [inline classification/feature_generation/feature_utilities.py]: Would you mind adding an extra sentence of description for how this differs from the above similarly named fn?

File: classification/feature_generation/feature_utilities.py [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+    num_features_inc_timestamp = series.shape[1]
+    return (np.empty(
+        [0, 1, window_size, num_features_inc_timestamp - 1],
+        dtype=np.float32), np.empty(
+            shape=[0, window_size], dtype=np.int32), np.empty(
+                shape=[0, 2], dtype=np.int32), np.empty(
+                    shape=[0], dtype=str))     
+
+
+def cook_features(features, id_):
```
Comment (d40cht) [inline classification/feature_generation/feature_utilities.py]: Description of what this does? :-)

File: classification/feature_generation/feature_utilities.py [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+        time_bounds = np.array([start_time, end_time], dtype=np.int32)
+        assert len(cropped) == window_size
+        slices.append(cook_features(cropped, id_))
+    if slices == []:
+        return empty_data(window_size, input_series)
+    return zip(*slices)
+
+
+
+def process_fixed_window_features(random_state, features, id_,
```
Comment (d40cht) [inline classification/feature_generation/feature_utilities.py]: Any chance of a comment explaining what this is for?

File: classification/feature_generation/file_iterator.py [hunk 11]
```diff
@@ -167,7 +163,7 @@ def all_fixed_window_feature_file_iterator(filenames, deserializer,
           2. A tensor of the timestamps for each feature point of dimension
              [n, window_size].
           3. A tensor of the timebounds for the slices, of dimension [n, 2].
-          4. A tensor of the mmsis of each vessel of dimension [n].
+          4. A tensor of the ids of each vessel of dimension [n].
```
Comment (d40cht) [inline classification/feature_generation/file_iterator.py]: Is the s/mmsi/id to become more generic and enable VMS etc?

File: classification/feature_generation/file_iterator_test.py [hunk 12]
```diff
@@ -18,36 +19,17 @@ def to_stamp(x):
     return (x - epoch).total_seconds()
 
 
-def test_file_iterator_2():
-    path = "gs://world-fishing-827/data-production/classification/release-0.1.2/pipeline/output/features/251822362.tfrecord"
```
Comment (d40cht) [inline classification/feature_generation/file_iterator_test.py]: Are we OK with tests not being hermetic and requiring a pull of data from GCS? Is this input data small enough it could be checked in to github?

File: classification/feature_generation/fishing_feature_generation.py (hunk 2 above)
Comment (d40cht) [reply classification/feature_generation/fishing_feature_generation.py]: Bird life style feature generation? Now I'm intrigued...

File: classification/models/fishing_detection.py [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+    FishingLocalizationObjectiveCrossEntropy, TrainNetInfo)
+from classification.feature_generation import fishing_feature_generation
+import logging
+import math
+import numpy as np
+import os
+
+import tensorflow as tf
+
+approx_means = [ 6.9139094e+00,  4.3506036e+00,  6.1346573e-01,  5.7922113e-01,
```
Comment (d40cht) [inline classification/models/fishing_detection.py]: Any chance of a comment here? I assume they are per-feature mean/stds for input normalisation? Also be nice to know how you estimated them.

File: classification/models/layers.py [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+        for _ in range(sub_layers - 1):
+            onet = ly.conv1d(onet, sub_filters, 1, activation=None, use_bias=False)
+            onet = ly.batch_normalization(onet, training=training, virtual_batch_size=virtual_batch_size)
+            onet = tf.nn.relu(onet)
+        onet = ly.conv1d(onet, sub_filters, 1, activation=tf.nn.relu)
+
+        # n = int(onet.get_shape().dims[1])
+        # onet = ly.average_pooling1d(onet, n, n)
+        onet = ly.flatten(onet)
+        #
```
Comment (d40cht) [inline classification/models/layers.py]: Tidy out commented lines?

File: classification/models/objectives.py [hunk 15]
```diff
@@ -0,0 +1,388 @@
...
+        except:
+            logging.warning("Problem creating eval_metrics in {}".format(self))
+            return {}
+        for k, v in eval_metrics.items():
+            tf.summary.scalar(k, v[1])
+        return eval_metrics
+
+
+
+class RegressionObjective(ObjectiveBase):
```
Comment (d40cht) [inline classification/models/objectives.py]: I'm not sure if you still have the regression heads for length, tonnage etc. But if you do, I've had a lot of recent success with quantising regression outputs, using softmax and then training with cross-entropy (having quantised and then one-hot encoded the label). Then using the tenth to ninetieth (for instance) percentile probability gap gives a nice estimate of model uncertainty.

File: classification/models/vessel_characterization.py [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+                XOrNan('tonnage'),
+                metrics=metrics,
+                loss_weight=0.1),
+            LogRegressionObjectiveMAE(
+                'engine_power',
+                'Vessel-engine-Power',
+                XOrNan('engine_power'),
+                metrics=metrics,
+                loss_weight=0.1),
+            LogRegressionObjectiveMAE(
```
Comment (d40cht) [inline classification/models/vessel_characterization.py]: Interesting. How does this perform?

File: classification/feature_generation/fishing_feature_generation.py (hunk 2 above)
Comment (bitsofbits) [reply classification/feature_generation/fishing_feature_generation.py]: @d40cht we put together a model for bird life that predicts longline sets (setting only, ignores hauling), that works quite well. It used time points interpolated onto a constant grid (and fewer features, although that's not the key point).  We'd like to move to interpolating onto a constant grid in the nearish future.

File: classification/feature_generation/feature_utilities.py (hunk 9 above)
Comment (bitsofbits) [reply classification/feature_generation/feature_utilities.py]: It cooks the features so that they are no longer raw ;-)

(I will add something).

File: classification/models/vessel_characterization.py (hunk 16 above)
Comment (bitsofbits) [reply classification/models/vessel_characterization.py]: Not obviously better or worse, unfortunately.

File: classification/models/objectives.py (hunk 15 above)
Comment (bitsofbits) [reply classification/models/objectives.py]: @d40cht , interesting. I've toyed with that idea, but never got around to trying it. Do you ever end up with bimodal distributions? If you do, do you do anything clever about it?

File: classification/models/fishing_detection.py (hunk 13 above)
Comment (bitsofbits) [reply classification/models/fishing_detection.py]: Will do.

File: classification/models/layers.py (hunk 14 above)
Comment (bitsofbits) [reply classification/models/layers.py]: ✔️

File: classification/feature_generation/feature_utilities.py (hunk 10 above)
Comment (bitsofbits) [reply classification/feature_generation/feature_utilities.py]: ✔️
